ably-ios/ARTRealtimeChannel.m

```diff
@@ -371,8 +371,10 @@ - (void)onMessage:(ARTProtocolMessage *)message {
             NSError *error = nil;
             msg = [msg decodeWithEncoder:dataEncoder error:&error];
             if (error != nil) {
-                [self.logger error:@"ARTRealtimeChannel: error decoding data: %@", error];
-                [NSException raise:NSInvalidArgumentException format:@"ARTRealtimeChannel: error decoding data: %@", error];
+                ARTErrorInfo *errorInfo = [ARTErrorInfo wrap:(ARTErrorInfo *)error.userInfo[NSLocalizedFailureReasonErrorKey] prepend:@"Failed to decode data: "];
+                [self.logger error:@"%@", errorInfo.message];
+                _errorReason = errorInfo;
+                [self emit:self.state with:errorInfo];
```

Comment:
I see, so apart from the channel states we have an "error" event that is emitted through `RealtimeChannel.on`, right?

I can see two alternatives: create a new enum `ChannelEvent`, with all the variants in `ChannelState` plus an "error" variant; or a more convoluted class `ChannelState` with two properties: `newState: ChannelState?` and `error: ErrorInfo?`, which are only populated when appropriate.